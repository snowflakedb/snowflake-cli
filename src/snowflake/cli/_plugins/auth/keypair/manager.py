from enum import Enum
from typing import Dict, List, Optional, Tuple

from click import ClickException
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric import rsa
from snowflake.cli._plugins.object.manager import ObjectManager
from snowflake.cli.api import exceptions
from snowflake.cli.api.cli_global_context import (
    _CliGlobalContextAccess,
    get_cli_context,
)
from snowflake.cli.api.config import (
    connection_exists,
    get_connection_dict,
    set_config_value,
)
from snowflake.cli.api.console import cli_console
from snowflake.cli.api.constants import ObjectType
from snowflake.cli.api.identifiers import FQN
from snowflake.cli.api.secret import SecretType
from snowflake.cli.api.secure_path import SecurePath
from snowflake.cli.api.sql_execution import SqlExecutionMixin
from snowflake.connector import DictCursor
from snowflake.connector.cursor import SnowflakeCursor


class PublicKeyProperty(Enum):
    RSA_PUBLIC_KEY = "RSA_PUBLIC_KEY"
    RSA_PUBLIC_KEY_2 = "RSA_PUBLIC_KEY_2"


class AuthManager(SqlExecutionMixin):
    def setup(
        self,
        connection_name: Optional[str],
        key_length: int,
        output_path: SecurePath,
        private_key_passphrase: SecretType,
    ):
        # When the user provide new connection name
        if connection_name and connection_exists(connection_name):
            raise ClickException(
                f"Connection with name {connection_name} already exists."
            )

        cli_context = get_cli_context()
        # When the use not provide connection name, so we overwrite the current connection
        if not connection_name:
            connection_name = cli_context.connection_context.connection_name

        key_name = AuthManager._get_free_key_name(output_path, connection_name)  # type: ignore[arg-type]
        self._generate_key_pair_and_set_public_key(
            user=cli_context.connection.user,
            key_length=key_length,
            output_path=output_path,
            key_name=key_name,  # type: ignore[arg-type]
            private_key_passphrase=private_key_passphrase,
        )

        self._create_or_update_connection(
            current_connection=cli_context.connection_context.connection_name,
            connection_name=connection_name,  # type: ignore[arg-type]
            private_key_path=self._get_private_key_path(
                output_path=output_path, key_name=key_name  # type: ignore[arg-type]
            ),
        )

    def rotate(
        self,
        key_length: int,
        output_path: SecurePath,
        private_key_passphrase: SecretType,
    ):
        cli_context = get_cli_context()
        connection_name = cli_context.connection_context.connection_name

        self._ensure_connection_has_private_key(
            cli_context.connection_context.connection_name
        )

        public_key, public_key_2 = self._get_public_keys()

        if not public_key and not public_key_2:
            raise ClickException("No public key found. Use the setup command first.")

        if public_key_2:
            self.set_public_key(
                cli_context.connection.user,
                PublicKeyProperty.RSA_PUBLIC_KEY,
                public_key_2,
            )

        key_name = AuthManager._get_free_key_name(output_path, connection_name)
        public_key = self._generate_keys_and_return_public_key(
            key_length=key_length,
            output_path=output_path,
            key_name=key_name,
            private_key_passphrase=private_key_passphrase,
        )
        self.set_public_key(
            cli_context.connection.user, PublicKeyProperty.RSA_PUBLIC_KEY_2, public_key
        )
        self._create_or_update_connection(
            current_connection=cli_context.connection_context.connection_name,
            connection_name=connection_name,
            private_key_path=self._get_private_key_path(
                output_path=output_path, key_name=key_name
            ),
        )

    def _generate_key_pair_and_set_public_key(
        self,
        user: str,
        key_length: int,
        output_path: SecurePath,
        key_name: str,
        private_key_passphrase: SecretType,
    ):
        public_key_exists, public_key_2_exists = self._get_public_keys()

        if public_key_exists or public_key_2_exists:
            raise exceptions.CouldNotSetKeyPairError()

        if not output_path.exists():
            output_path.mkdir(parents=True)

        public_key = self._generate_keys_and_return_public_key(
            key_length=key_length,
            output_path=output_path,
            key_name=key_name,  # type: ignore[arg-type]
            private_key_passphrase=private_key_passphrase,
        )
        self.set_public_key(user, PublicKeyProperty.RSA_PUBLIC_KEY, public_key)

    def list_keys(self) -> List[Dict]:
        key_properties = [
            "RSA_PUBLIC_KEY",
            "RSA_PUBLIC_KEY_FP",
            "RSA_PUBLIC_KEY_LAST_SET_TIME",
            "RSA_PUBLIC_KEY_2",
            "RSA_PUBLIC_KEY_2_FP",
            "RSA_PUBLIC_KEY_2_LAST_SET_TIME",
        ]
        cursor = ObjectManager(connection=self._conn).describe(
            object_type=ObjectType.USER.value.sf_name,
            fqn=FQN.from_string(self._conn.user),
            cursor_class=DictCursor,
        )
        only_public_key_properties = [
            p for p in cursor.fetchall() if p.get("property") in key_properties
        ]
        return only_public_key_properties

    def set_public_key(
        self, user: str, public_key_property: PublicKeyProperty, public_key: str
    ) -> SnowflakeCursor:
        return self.execute_query(
            f"ALTER USER {user} SET {public_key_property.value}='{public_key}'"
        )

    def remove_public_key(
        self, public_key_property: PublicKeyProperty
    ) -> SnowflakeCursor:
        cli_context = get_cli_context()
        return self.execute_query(
            f"ALTER USER {cli_context.connection.user} UNSET {public_key_property.value}"
        )

    def status(self):
        cli_context = get_cli_context()
        self._ensure_connection_has_private_key(
            cli_context.connection_context.connection_name
        )
        cli_console.step("Private key set for connection - OK")
        self._check_connection(cli_context)
        cli_console.step("Test connection - OK")

    def extend_connection_add(
        self,
        connection_name: str,
        connection_options: Dict,
        key_length: int,
        output_path: SecurePath,
        private_key_passphrase: SecretType,
    ) -> Dict:
        key_name = AuthManager._get_free_key_name(output_path, connection_name)

        self._generate_key_pair_and_set_public_key(
            user=connection_options["user"],
            key_length=key_length,
            output_path=output_path,
            key_name=key_name,
            private_key_passphrase=private_key_passphrase,
        )

        connection_options["authenticator"] = "SNOWFLAKE_JWT"
        connection_options["private_key_file"] = str(
            self._get_private_key_path(output_path=output_path, key_name=key_name).path
        )
        if connection_options.get("password"):
            del connection_options["password"]

        return connection_options

    @staticmethod
    def _ensure_connection_has_private_key(connection_name: str) -> None:
        connection = get_connection_dict(connection_name)
        if not connection.get("private_key_file") and not connection.get(
            "private_key_path"
        ):
            raise ClickException(
                f"The private key is not set in {connection_name} connection."
            )

    @staticmethod
    def _check_connection(cli_context: _CliGlobalContextAccess) -> None:
        cli_context.connection

    def _get_public_keys(self) -> Tuple[str, str]:
        keys = self.list_keys()
        public_key = ""
        public_key_2 = ""
        for p in keys:
            if (
                p.get("property") == PublicKeyProperty.RSA_PUBLIC_KEY.value
                and p.get("value") != "null"
            ):
                public_key = p.get("value")  # type: ignore
            if (
                p.get("property") == PublicKeyProperty.RSA_PUBLIC_KEY_2.value
                and p.get("value") != "null"
            ):
                public_key_2 = p.get("value")  # type: ignore
        return public_key, public_key_2

    @staticmethod
    def _generate_keys_and_return_public_key(
        key_length: int,
        output_path: SecurePath,
        key_name: str,
        private_key_passphrase: SecretType,
    ) -> str:
        private_key = rsa.generate_private_key(
            public_exponent=65537,
            key_size=key_length,
        )

        if private_key_passphrase:
            pem = SecretType(
                private_key.private_bytes(
                    encoding=serialization.Encoding.PEM,
                    format=serialization.PrivateFormat.PKCS8,
                    encryption_algorithm=serialization.BestAvailableEncryption(
                        private_key_passphrase.value.encode("utf-8")
                    ),
                )
            )
            cli_console.message(
                "Set the `PRIVATE_KEY_PASSPHRASE` environment variable before using the connection."
            )
        else:
            pem = SecretType(
                private_key.private_bytes(
                    encoding=serialization.Encoding.PEM,
                    format=serialization.PrivateFormat.PKCS8,
                    encryption_algorithm=serialization.NoEncryption(),
                )
            )

        with AuthManager._get_private_key_path(output_path, key_name).open(
            mode="wb"
        ) as file:
            file.write(pem.value)

        public_key = private_key.public_key()
        public_pem = public_key.public_bytes(
            encoding=serialization.Encoding.PEM,
            format=serialization.PublicFormat.SubjectPublicKeyInfo,
        )
        with AuthManager._get_public_key_path(output_path, key_name).open(
            mode="wb"
        ) as file:
            file.write(public_pem)

        return public_pem.decode("utf-8")

    @staticmethod
    def _get_free_key_name(output_path: SecurePath, key_name: str) -> str:
        new_private_key = f"{key_name}.p8"
        new_public_key = f"{key_name}.pub"
        new_key_name = key_name
        counter = 1

        while (
            (output_path / new_private_key).exists()
            and (output_path / new_public_key).exists()
            and counter <= 100
        ):
            new_key_name = f"{key_name}_{counter}"
            new_private_key = f"{new_key_name}.p8"
            new_public_key = f"{new_key_name}.pub"
            counter += 1

        if counter == 100:
            raise ClickException(
                "Too many key pairs with the same name in the output directory."
            )

        return new_key_name

    @staticmethod
    def _get_private_key_path(output_path: SecurePath, key_name: str) -> SecurePath:
        return (output_path / f"{key_name}.p8").resolve()

    @staticmethod
    def _get_public_key_path(output_path: SecurePath, key_name: str) -> SecurePath:
        return (output_path / f"{key_name}.pub").resolve()

    @staticmethod
    def _create_or_update_connection(
        current_connection: Optional[str],
        connection_name: str,
        private_key_path: SecurePath,
    ):
        connection = get_connection_dict(current_connection)
        connection.pop("password", None)
        connection["authenticator"] = "SNOWFLAKE_JWT"
        connection["private_key_file"] = str(private_key_path.path)

        set_config_value(["connections", connection_name], value=connection)
