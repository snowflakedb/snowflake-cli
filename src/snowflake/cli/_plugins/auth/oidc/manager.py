# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import logging
from typing import TypeAlias

from snowflake.cli._app.auth.errors import OidcProviderError
from snowflake.cli._app.auth.oidc_providers import (
    OidcProviderType,
    OidcProviderTypeWithAuto,
    auto_detect_oidc_provider,
    get_active_oidc_provider,
)
from snowflake.cli.api.exceptions import CliError

logger = logging.getLogger(__name__)


Providers: TypeAlias = OidcProviderType | OidcProviderTypeWithAuto


class OidcManager:
    """
    Manages OIDC authentication.

    This class provides methods to read OIDC configurations for authentication.
    """

    def read_token(
        self,
        provider_type: Providers = OidcProviderTypeWithAuto.AUTO,
    ) -> str:
        """
        Reads OIDC token based on the specified provider type.

        Args:
            provider_type: Type of provider to read token from ("auto" for auto-detection)

        Returns:
            Token string or provider information

        Raises:
            CliError: If token reading fails
        """
        logger.info("Reading OIDC token with provider type: %s", provider_type)

        try:
            if provider_type == OidcProviderTypeWithAuto.AUTO:
                provider = auto_detect_oidc_provider()
            else:
                provider = get_active_oidc_provider(provider_type.value)
            return provider.get_token()
        except OidcProviderError as e:
            logger.error("OIDC provider error: %s", str(e))
            raise CliError(str(e))
