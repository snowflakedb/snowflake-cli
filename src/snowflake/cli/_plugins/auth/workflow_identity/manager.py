# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import Optional

from click import ClickException
from snowflake.cli.api.cli_global_context import get_cli_context
from snowflake.cli.api.console import cli_console
from snowflake.cli.api.sql_execution import SqlExecutionMixin
from snowflake.connector.cursor import SnowflakeCursor


class WorkflowIdentityManager(SqlExecutionMixin):
    """
    Manager for GitHub workflow identity federation authentication.
    """

    def setup(self, github_repository: str) -> None:
        """
        Sets up GitHub workflow identity federation for the specified repository.

        Args:
            github_repository: GitHub repository in format 'owner/repo'
        """
        if not self._validate_github_repository_format(github_repository):
            raise ClickException(
                "Invalid GitHub repository format. Expected format: 'owner/repo'"
            )

        cli_context = get_cli_context()
        user = cli_context.connection.user

        # Create or update the security integration for GitHub workflow identity federation
        self._create_security_integration(github_repository)
        cli_console.step(f"Created security integration for {github_repository}")

        # Set the workflow identity property for the user
        self._set_workflow_identity_property(user, github_repository)
        cli_console.step(f"Set workflow identity property for user {user}")

    def status(self) -> str:
        """
        Returns the status of GitHub workflow identity federation configuration.

        Returns:
            Status message indicating if workflow identity federation is configured
        """
        cli_context = get_cli_context()
        user = cli_context.connection.user

        try:
            # Check if workflow identity is configured for the user
            workflow_identity = self._get_workflow_identity_property(user)
            if workflow_identity:
                return f"Configured for repository: {workflow_identity}"
            else:
                return "Not configured"
        except Exception as e:
            return f"Error checking status: {str(e)}"

    def remove(self) -> SnowflakeCursor:
        """
        Removes the GitHub workflow identity federation configuration.

        Returns:
            SnowflakeCursor with the result of the operation
        """
        cli_context = get_cli_context()
        user = cli_context.connection.user

        return self.execute_query(f"ALTER USER {user} UNSET GITHUB_WORKFLOW_IDENTITY")

    def _validate_github_repository_format(self, repository: str) -> bool:
        """
        Validates that the GitHub repository is in the correct format (owner/repo).

        Args:
            repository: The repository string to validate

        Returns:
            True if format is valid, False otherwise
        """
        parts = repository.split("/")
        return len(parts) == 2 and all(part.strip() for part in parts)

    def _create_security_integration(self, github_repository: str) -> SnowflakeCursor:
        """
        Creates or replaces a security integration for GitHub workflow identity federation.

        Args:
            github_repository: GitHub repository in format 'owner/repo'

        Returns:
            SnowflakeCursor with the result of the operation
        """
        integration_name = f"GITHUB_WIF_{github_repository.replace('/', '_').upper()}"

        sql = f"""
        CREATE OR REPLACE SECURITY INTEGRATION {integration_name}
        TYPE = OAUTH
        OAUTH_CLIENT = GITHUB_ACTIONS
        OAUTH_ALLOWED_AUTHORIZATION_ENDPOINTS = ('https://token.actions.githubusercontent.com')
        OAUTH_ALLOWED_TOKEN_ENDPOINTS = ('https://token.actions.githubusercontent.com')
        OAUTH_ALLOWED_SCOPES = ('repository:read')
        OAUTH_ISSUER = 'https://token.actions.githubusercontent.com'
        OAUTH_AUDIENCE = '{github_repository}'
        """

        return self.execute_query(sql)

    def _set_workflow_identity_property(
        self, user: str, github_repository: str
    ) -> SnowflakeCursor:
        """
        Sets the GitHub workflow identity property for a user.

        Args:
            user: The Snowflake user name
            github_repository: GitHub repository in format 'owner/repo'

        Returns:
            SnowflakeCursor with the result of the operation
        """
        return self.execute_query(
            f"ALTER USER {user} SET GITHUB_WORKFLOW_IDENTITY = '{github_repository}'"
        )

    def _get_workflow_identity_property(self, user: str) -> Optional[str]:
        """
        Gets the GitHub workflow identity property for a user.

        Args:
            user: The Snowflake user name

        Returns:
            The workflow identity property value if set, None otherwise
        """
        cursor = self.execute_query(f"DESCRIBE USER {user}")
        properties = cursor.fetchall()

        for prop in properties:
            if prop[0] == "GITHUB_WORKFLOW_IDENTITY" and prop[1]:
                return prop[1]

        return None
