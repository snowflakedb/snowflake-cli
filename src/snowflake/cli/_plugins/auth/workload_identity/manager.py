# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import logging

from snowflake.cli._plugins.auth.workload_identity.oidc_providers import (
    auto_detect_oidc_provider,
    get_oidc_provider,
    list_oidc_providers,
)
from snowflake.cli.api.exceptions import CliError
from snowflake.cli.api.sql_execution import SqlExecutionMixin

logger = logging.getLogger(__name__)


class WorkloadIdentityManager(SqlExecutionMixin):
    """
    Manager for GitHub workload identity federation authentication.
    """

    def setup(self, github_repository: str) -> str:
        """
        Sets up GitHub workload identity federation for the specified repository.

        Args:
            github_repository: GitHub repository in format 'owner/repo'

        Returns:
            Success message string

        Raises:
            NotImplementedError: Setup functionality is not yet implemented
        """
        logger.info(
            "Attempting to set up GitHub workload identity federation for repository: %s",
            github_repository,
        )
        logger.warning(
            "GitHub workload identity federation setup is not yet implemented"
        )
        raise NotImplementedError(
            "GitHub workload identity federation setup is not yet implemented"
        )

    def read(self, provider_type: str) -> str:
        """
        Reads OIDC token based on the specified type.

        Args:
            provider_type: Type of provider ('auto' for auto-detection or specific provider name)

        Returns:
            The OIDC token string

        Raises:
            CliError: If token cannot be retrieved or provider is not available
        """
        logger.info("Reading OIDC token with provider type: %s", provider_type)

        if provider_type == "auto":
            logger.debug("Using auto-detection for OIDC token")
            return self._read_auto_detect_token()
        else:
            logger.debug("Using specific provider for OIDC token: %s", provider_type)
            return self._read_specific_token(provider_type)

    def _read_auto_detect_token(self) -> str:
        """
        Auto-detects and reads OIDC token from available providers.

        Returns:
            The OIDC token string

        Raises:
            CliError: If no providers are available or token cannot be retrieved
        """
        logger.debug("Starting auto-detection of OIDC providers")
        provider = auto_detect_oidc_provider()

        if not provider:
            logger.warning("No OIDC provider detected in current environment")
            available_providers = list_oidc_providers()
            logger.debug("Available providers: %s", available_providers)

            if available_providers:
                providers_list = ", ".join(available_providers)
                error_msg = (
                    "No OIDC provider detected in current environment. "
                    "Available providers: %s. "
                    "Use --type <provider> to specify a provider explicitly."
                ) % providers_list
                logger.error(error_msg)
                raise CliError(error_msg)
            else:
                error_msg = "No OIDC providers are registered."
                logger.error(error_msg)
                raise CliError(error_msg)

        logger.info("Auto-detected OIDC provider: %s", provider.provider_name)

        try:
            logger.debug("Retrieving token from provider: %s", provider.provider_name)
            token = provider.get_token()

            logger.debug(
                "Retrieving token info from provider: %s", provider.provider_name
            )
            token_info = provider.get_token_info()

            info_str = "Provider: %s" % provider.provider_name
            if token_info:
                info_details = ", ".join(
                    ["%s: %s" % (k, v) for k, v in token_info.items()]
                )
                info_str += " (%s)" % info_details

            logger.info(
                "Successfully retrieved OIDC token via auto-detection: %s", info_str
            )
            return token
        except Exception as e:
            error_msg = "Failed to retrieve token from %s: %s" % (
                provider.provider_name,
                str(e),
            )
            logger.error(error_msg)
            raise CliError(error_msg)

    def _read_specific_token(self, provider_name: str) -> str:
        """
        Reads OIDC token from a specific provider.

        Args:
            provider_name: Name of the provider to use

        Returns:
            The OIDC token string

        Raises:
            CliError: If provider is unknown, unavailable, or token cannot be retrieved
        """
        logger.debug("Reading OIDC token from specific provider: %s", provider_name)
        provider = get_oidc_provider(provider_name)

        if not provider:
            logger.warning("Provider '%s' not found", provider_name)
            available_providers = list_oidc_providers()
            logger.debug("Available providers: %s", available_providers)
            providers_list = ", ".join(available_providers)
            error_msg = "Unknown provider '%s'. Available providers: %s" % (
                provider_name,
                providers_list,
            )
            logger.error(error_msg)
            raise CliError(error_msg)

        logger.debug("Checking availability of provider: %s", provider_name)
        if not provider.is_available:
            error_msg = (
                "Provider '%s' is not available in the current environment."
                % provider_name
            )
            logger.error(error_msg)
            raise CliError(error_msg)

        logger.info("Using provider: %s", provider_name)

        try:
            logger.debug("Retrieving token from provider: %s", provider_name)
            token = provider.get_token()

            logger.debug("Retrieving token info from provider: %s", provider_name)
            token_info = provider.get_token_info()

            info_str = "Provider: %s" % provider.provider_name
            if token_info:
                info_details = ", ".join(
                    ["%s: %s" % (k, v) for k, v in token_info.items()]
                )
                info_str += " (%s)" % info_details

            logger.info(
                "Successfully retrieved OIDC token from %s: %s", provider_name, info_str
            )
            return token
        except Exception as e:
            error_msg = "Failed to retrieve token from %s: %s" % (provider_name, str(e))
            logger.error(error_msg)
            raise CliError(error_msg)
