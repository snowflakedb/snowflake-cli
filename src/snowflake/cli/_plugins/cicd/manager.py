import enum
from typing import List, Optional, Type

from snowflake.cli.api.secure_path import SecurePath


class CIProviderChoices(str, enum.Enum):
    GITHUB = "GITHUB"
    GITLAB = "GITLAB"


class CIProvider:
    NAME: str

    @classmethod
    def cleanup(cls, root: SecurePath) -> None:
        raise NotImplementedError()

    @classmethod
    def from_choice(cls, choice: CIProviderChoices) -> "CIProvider":
        return {
            GithubProvider.NAME: GithubProvider,
            GitLabProvider.NAME: GitLabProvider,
        }[choice.name]()

    @classmethod
    def all(cls) -> List[Type["CIProvider"]]:  # noqa: A003
        return [GithubProvider, GitLabProvider]

    def has_template(self, root_dir: SecurePath) -> bool:
        raise NotImplementedError()

    def copy(self, source: SecurePath, destination: SecurePath):
        raise NotImplementedError()


class GithubProvider(CIProvider):
    NAME = CIProviderChoices.GITHUB.name

    @classmethod
    def cleanup(cls, root_dir: SecurePath):
        (root_dir / ".github").rmdir(recursive=True)

    def has_template(self, root_dir: SecurePath) -> bool:
        return (root_dir / ".github/workflows").exists()

    def copy(self, source: SecurePath, destination: SecurePath) -> None:
        (source / ".github").copy(destination.path, dirs_exist_ok=True)


class GitLabProvider(CIProvider):
    NAME = CIProviderChoices.GITLAB.name

    @classmethod
    def cleanup(cls, root_dir: SecurePath):
        (root_dir / ".gitlab-ci.yml").unlink(missing_ok=True)

    def has_template(self, root_dir: SecurePath) -> bool:
        return (root_dir / ".gitlab-ci.yml").exists()

    def copy(self, source: SecurePath, destination: SecurePath) -> None:
        if (destination / ".gitlab-ci.yml").exists():
            (destination / ".gitlab-ci.yml").unlink()
        (source / ".gitlab-ci.yml").move(destination.path)


class CIProviderManager:
    @staticmethod
    def project_post_gen_cleanup(
        selected_provider: Optional[CIProvider], template_root: SecurePath
    ):
        for provider_cls in CIProvider.all():
            if selected_provider and not isinstance(selected_provider, provider_cls):
                provider_cls.cleanup(template_root)
