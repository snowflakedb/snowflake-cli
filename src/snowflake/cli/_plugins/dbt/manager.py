# Copyright (c) 2025 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

from pathlib import Path
from typing import Optional

import yaml
from click import ClickException
from snowflake.cli._plugins.stage.manager import StageManager
from snowflake.cli.api.console import cli_console
from snowflake.cli.api.feature_flags import FeatureFlag
from snowflake.cli.api.identifiers import FQN
from snowflake.cli.api.sql_execution import SqlExecutionMixin
from snowflake.connector.cursor import SnowflakeCursor


class StdoutExecutionMixin(SqlExecutionMixin):
    def execute_query(self, query, **kwargs):
        if FeatureFlag.ENABLE_DBT_POC.is_enabled():
            from unittest.mock import MagicMock

            cli_console.message(f"Sending query: {query}")
            mock_cursor = MagicMock()
            mock_cursor.description = []
            return mock_cursor
        return super().execute_query(query, **kwargs)


class DBTManager(StdoutExecutionMixin):
    def list(self) -> SnowflakeCursor:  # noqa: A003
        query = "SHOW DBT PROJECT"
        return self.execute_query(query)

    def deploy(
        self,
        path: Path,
        name: FQN,
        dbt_version: Optional[str],
        dbt_adapter_version: str,
        force: bool,
    ) -> SnowflakeCursor:
        # TODO: what to do with force?
        dbt_project_path = path.joinpath("dbt_project.yml")
        if not dbt_project_path.exists():
            raise ClickException(f"dbt_project.yml does not exist in provided path.")

        if dbt_version is None:
            with dbt_project_path.open() as fd:
                dbt_project_config = yaml.safe_load(fd)
                try:
                    dbt_version = dbt_project_config["version"]
                except (KeyError, TypeError):
                    raise ClickException(
                        f"dbt-version was not provided and is not available in dbt_project.yml"
                    )

        with cli_console.phase("Creating temporary stage"):
            stage_manager = StageManager()
            stage_fqn = FQN.from_string(f"dbt_{name}_stage").using_context()
            stage_name = stage_manager.get_standard_stage_prefix(stage_fqn)
            stage_manager.create(stage_fqn, temporary=True)

        with cli_console.phase("Copying project files to stage"):
            results = list(stage_manager.put_recursive(path, stage_name))
            cli_console.step(f"Copied {len(results)} files")

        with cli_console.phase("Creating DBT project"):
            staged_dbt_project_path = self._get_dbt_project_stage_path(stage_name)
            query = f"""CREATE OR REPLACE DBT PROJECT {name}
FROM {stage_name} MAIN_FILE='{staged_dbt_project_path}'
DBT_VERSION='{dbt_version}' DBT_ADAPTER_VERSION='{dbt_adapter_version}'"""
            return self.execute_query(query)

    def execute(self, dbt_command: str, name: str, *dbt_cli_args):
        query = f"EXECUTE DBT PROJECT {name} {dbt_command}"
        if dbt_cli_args:
            query += " " + " ".join([arg for arg in dbt_cli_args])
        return self.execute_query(query)

    @staticmethod
    def _get_dbt_project_stage_path(stage_name):
        return "/".join([stage_name, "dbt_project.yml"])
