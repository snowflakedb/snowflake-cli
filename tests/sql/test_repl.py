from unittest import mock

import pytest
from snowflake.cli._plugins.sql.manager import SqlManager
from snowflake.cli._plugins.sql.repl import Repl


@pytest.fixture(name="repl")
def make_repl(mock_cursor):
    mocked_cursor = [
        mock_cursor(
            rows=[("1",)],
            columns=["1"],
        ),
    ]

    with mock.patch.object(SqlManager, "_execute_string", return_value=mocked_cursor):
        manager = SqlManager()
        repl = Repl(manager)
        yield repl


def test_repl_input_handling(repl, capsys, os_agnostic_snapshot):
    user_inputs = iter(("select 1;", "exit", "y"))

    with mock.patch.object(
        repl.session,
        "prompt",
        side_effect=user_inputs,
    ):
        repl.run()

    output = capsys.readouterr().out
    os_agnostic_snapshot.assert_match(output)


@pytest.mark.parametrize(
    "user_inputs",
    (
        pytest.param(("exit", "y"), id="exit"),
        pytest.param(("quit", "y"), id="quit"),
        pytest.param(("exit", "n", "exit", "y"), id="hesistate on exit"),
        pytest.param(("quit", "n", "quit", "y"), id="hesistate on quit"),
    ),
)
def test_exit_sequence(user_inputs, repl, os_agnostic_snapshot, capsys):
    user_inputs = iter(user_inputs)

    with mock.patch.object(
        repl.session,
        "prompt",
        side_effect=user_inputs,
    ):
        repl.run()

    output = capsys.readouterr().out
    os_agnostic_snapshot.assert_match(output)


def test_repl_full_app(runner, os_agnostic_snapshot, mock_cursor):
    user_inputs = iter(("exit", "y"))
    mocked_cursor = [
        mock_cursor(
            rows=[("1",)],
            columns=("1",),
        ),
    ]

    repl_prompt = "snowflake.cli._plugins.sql.repl.PromptSession"
    repl_execute = "snowflake.cli._plugins.sql.repl.Repl._execute"

    with mock.patch(repl_prompt) as mock_prompt:
        mock_instance = mock.MagicMock()
        mock_instance.prompt.side_effect = user_inputs
        mock_prompt.return_value = mock_instance

        with mock.patch(repl_execute, return_value=mocked_cursor):
            result = runner.invoke(("sql",))
            assert result.exit_code == 0
            os_agnostic_snapshot.assert_match(result.output)
