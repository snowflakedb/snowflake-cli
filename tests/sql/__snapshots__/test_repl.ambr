# serializer version: 1
# name: test_exit_sequence[exit]
  '''
  ╭──────────────────────────────────────────────────────────────────────────────╮
  │ Welcome to Snowflake-CLI REPL                                                │
  │ Type 'exit' or 'quit' to leave                                               │
  ╰──────────────────────────────────────────────────────────────────────────────╯
  
  Leaving REPL, bye ...
  
  '''
# ---
# name: test_exit_sequence[hesistate on exit]
  '''
  ╭──────────────────────────────────────────────────────────────────────────────╮
  │ Welcome to Snowflake-CLI REPL                                                │
  │ Type 'exit' or 'quit' to leave                                               │
  ╰──────────────────────────────────────────────────────────────────────────────╯
  
  Leaving REPL, bye ...
  
  '''
# ---
# name: test_exit_sequence[hesistate on quit]
  '''
  ╭──────────────────────────────────────────────────────────────────────────────╮
  │ Welcome to Snowflake-CLI REPL                                                │
  │ Type 'exit' or 'quit' to leave                                               │
  ╰──────────────────────────────────────────────────────────────────────────────╯
  
  Leaving REPL, bye ...
  
  '''
# ---
# name: test_exit_sequence[quit]
  '''
  ╭──────────────────────────────────────────────────────────────────────────────╮
  │ Welcome to Snowflake-CLI REPL                                                │
  │ Type 'exit' or 'quit' to leave                                               │
  ╰──────────────────────────────────────────────────────────────────────────────╯
  
  Leaving REPL, bye ...
  
  '''
# ---
# name: test_repl_full_app
  '''
  ╭──────────────────────────────────────────────────────────────────────────────╮
  │ Welcome to Snowflake-CLI REPL                                                │
  │ Type 'exit' or 'quit' to leave                                               │
  ╰──────────────────────────────────────────────────────────────────────────────╯
  
  Leaving REPL, bye ...
  
  
  '''
# ---
# name: test_repl_input_handling
  '''
  ╭──────────────────────────────────────────────────────────────────────────────╮
  │ Welcome to Snowflake-CLI REPL                                                │
  │ Type 'exit' or 'quit' to leave                                               │
  ╰──────────────────────────────────────────────────────────────────────────────╯
  +---+
  | 1 |
  |---|
  | 1 |
  +---+
  
  Leaving REPL, bye ...
  
  '''
# ---
